File: canSum.java
import java.util.Map;
import java.util.HashMap;

public class canSum{

    //Memoisation
    private Map<Integer, Boolean> memo = new HashMap<>();

    public boolean sum(int target, int[] numbers){

        if(memo.containsKey(target)){
            return memo.get(target);
        }

        if(target == 0){
            return true;
        }

        if(target < 0){
            return false;
        }

        for(int i=0; i<numbers.length;i++){
           int remainder = target - numbers[i];

           if(sum(remainder, numbers) == true){
                memo.put(target, true);
                return true;
           }
        }
        memo.put(target, false);
        return false;
    }

    public static void main(String[] args){
        canSum sum = new canSum();

        // Test Case: 1
        int[] numbers = {7, 14};
        System.out.print(sum.sum(300, numbers));
    }
}

File: canSum.py
def canSum(target: int, numbers: list, memo = {}) -> bool:
    """
    Solved recursively using memoisation

    Returns True if it's possible to get the target
    by summing a combination of digits in numbers(list)
    """
    # Memoisation
    if target in memo:
        return memo[target]
    if target == 0:
        return True
    
    if target < 0:
        return False

    for num in numbers:
        remainder = target - num
        if canSum(remainder, numbers):
            memo[target] = True
            return True
    memo[target] = False
    return False

# Test Case: 1
print(canSum(300, [7, 14]))